File: ruby/lib/steam/community/defense_grid/defense_grid_stats.rb
# This code is free software; you can redistribute it and/or modify it under
# the terms of the new BSD License.
#
# Copyright (c) 2009-2011, Sebastian Staudt

require 'steam/community/game_stats'

# This class represents the game statistics for a single user in Defense Grid:
# The Awakening
#
# @author Sebastian Staudt
class DefenseGridStats < GameStats

  # Returns the
  #
  # @return []
  attr_reader :bronze_medals

  # Returns the
  #
  # @return []
  attr_reader :damage_done

  # Returns the
  #
  # @return []
  attr_reader :damage_campaign

  # Returns the
  #
  # @return []
  attr_reader :damage_challenge

  # Returns the
  #
  # @return []
  attr_reader :encountered

  # Returns the
  #
  # @return []
  attr_reader :gold_medals

  # Returns the
  #
  # @return []
  attr_reader :heat_damage

  # Returns the
  #
  # @return []
  attr_reader :interest

  # Returns the
  #
  # @return []
  attr_reader :killed

  # Returns the
  #
  # @return []
  attr_reader :killed_campaign

  # Returns the
  #
  # @return []
  attr_reader :killed_challenge

  # Returns the
  #
  # @return []
  attr_reader :levels_played

  # Returns the
  #
  # @return []
  attr_reader :levels_played_campaign

  # Returns the
  #
  # @return []
  attr_reader :levels_played_challenge

  # Returns the
  #
  # @return []
  attr_reader :levels_won

  # Returns the
  #
  # @return []
  attr_reader :levels_won_campaign

  # Returns the
  #
  # @return []
  attr_reader :levels_won_challenge

  # Returns the
  #
  # @return []
  attr_reader :orbital_laser_fired

  # Returns the
  #
  # @return []
  attr_reader :orbital_laser_damage

  # Returns the
  #
  # @return []
  attr_reader :resources

  # Returns the
  #
  # @return []
  attr_reader :silver_medals

  # Returns the
  #
  # @return []
  attr_reader :time_played

  # Creates a `DefenseGridStats` instance by calling the super constructor with
  # the game name `'defensegrid:awakening'`
  #
  # @param [String, Fixnum] steam_id The custom URL or the 64bit Steam ID of
  #        the user
  def initialize(steam_id)
    super(steam_id, 'defensegrid:awakening')

    if public?
      general_data = @xml_data.elements['stats/general']

      @bronze_medals           = general_data.elements['bronze_medals_won/value'].text.to_i
      @silver_medals           = general_data.elements['silver_medals_won/value'].text.to_i
      @gold_medals             = general_data.elements['gold_medals_won/value'].text.to_i
      @levels_played           = general_data.elements['levels_played_total/value'].text.to_i
      @levels_played_campaign  = general_data.elements['levels_played_campaign/value'].text.to_i
      @levels_played_challenge = general_data.elements['levels_played_challenge/value'].text.to_i
      @levels_won              = general_data.elements['levels_won_total/value'].text.to_i
      @levels_won_campaign     = general_data.elements['levels_won_campaign/value'].text.to_i
      @levels_won_challenge    = general_data.elements['levels_won_challenge/value'].text.to_i
      @encountered             = general_data.elements['total_aliens_encountered/value'].text.to_i
      @killed                  = general_data.elements['total_aliens_killed/value'].text.to_i
      @killed_campaign         = general_data.elements['total_aliens_killed_campaign/value'].text.to_i
      @killed_challenge        = general_data.elements['total_aliens_killed_challenge/value'].text.to_i
      @resources               = general_data.elements['resources_recovered/value'].text.to_i
      @heat_damage             = general_data.elements['heatdamage/value'].text.to_f
      @time_played             = general_data.elements['time_played/value'].text.to_f
      @interest                = general_data.elements['interest_gained/value'].text.to_f
      @damage_done             = general_data.elements['tower_damage_total/value'].text.to_f
      @damage_campaign         = general_data.elements['tower_damage_total_campaign/value'].text.to_f
      @damage_challenge        = general_data.elements['tower_damage_total_challenge/value'].text.to_f
      @orbital_laser_fired     = @xml_data.elements['stats/orbitallaser/fired/value'].text.to_i
      @orbital_laser_damage    = @xml_data.elements['stats/orbitallaser/damage/value'].text.to_f
    end
  end

  # Returns stats about the aliens encountered by the player
  #
  # The Hash returned uses the names of the aliens as keys. Every value of the
  # Hash is an Array containing the number of aliens encountered as the first
  # element and the number of aliens killed as the second element.
  #
  # @return [Hash<String, Array<Fixnum>>] Stats about the aliens encountered
  def alien_stats
    return unless public?

    if @alien_stats.nil?
      alien_data = @xml_data.elements['stats/aliens']
      @alien_stats = {}
      aliens = %w{swarmer juggernaut crasher spire grunt bulwark drone manta dart
                  decoy rumbler seeker turtle walker racer stealth}

      aliens.each do |alien|
        @alien_stats[alien] = [
          alien_data.elements["#{alien}/encountered/value"].text.to_i,
          alien_data.elements["#{alien}/killed/value"].text.to_i
        ]
      end
    end

    @alien_stats
  end

  # Returns stats about the towers built by the player
  #
  # The Hash returned uses the names of the towers as keys. Every value of
  # the Hash is another Hash using the keys 1 to 3 for different tower levels.
  # The values of these Hash is an Array containing the number of towers built
  # as the first element and the damage dealt by this specific tower type as the
  # second element.
  #
  # The Command tower uses the resources gained as second element.
  # The Temporal tower doesn't have a second element.
  #
  # @return [Hash<String, Array<Fixnum>>] Stats about the towers built
  def tower_stats
    return unless public?

    if @tower_stats.nil?
      tower_data = @xml_data.elements['stats/towers']
      @tower_stats = {}
      towers = %w{cannon flak gun inferno laser meteor missile tesla}

      towers.each do |tower|
        @tower_stats[tower] = {}
        (1..3).each do |i|
          @tower_stats[tower][i] = [
            tower_data.elements["#{tower}[@level=#{i}]/built/value"].text.to_i,
            tower_data.elements["#{tower}[@level=#{i}]/damage/value"].text.to_f
          ]
        end
      end

      @tower_stats['command'] = {}
      (1..3).each do |i|
        @tower_stats['command'][i] = [
          tower_data.elements["command[@level=#{i}]/built/value"].text.to_i,
          tower_data.elements["command[@level=#{i}]/resource/value"].text.to_f
        ]
      end

      @tower_stats['temporal'] = {}
      (1..3).each do |i|
        @tower_stats['temporal'][i] = [
          tower_data.elements["temporal[@level=#{i}]/built/value"].text.to_i,
        ]
      end
    end

    @tower_stats
  end

end

File: php/lib/steam/community/defense_grid/DefenseGridStats.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under the
 * terms of the new BSD License.
 *
 * Copyright (c) 2009, Sebastian Staudt
 *
 * @author     Sebastian Staudt
 * @license    http://www.opensource.org/licenses/bsd-license.php New BSD License
 * @package    Steam Condenser (PHP)
 * @subpackage Steam Community
 */

require_once STEAM_CONDENSER_PATH . 'steam/community/GameStats.php';

/**
 * The DefenseGridStats class represents the game statistics for a single user
 * in Defense Grid: The Awakening
 *
 * @package Steam Condenser (PHP)
 * @subpackage Steam Community
 */
class DefenseGridStats extends GameStats {
    /**
     * @var array
     */
    public $towerStats;
    /**
     * @var array
     */
    public $alienStats;

    /**
     * @var int
     */
    public $goldMedals;

    /**
     *
     * @var int
     */
    public $silverMedals;

    /**
     *
     * @var int
     */
    public $bronzeMedals;

    /**
     * Creates a DefenseGridStats object by calling the super constructor with
     * the game name "defensegrid:awakening"
     *
     * @param String $steamId
     */
    public function __construct($steamId) {
        parent::__construct($steamId, 'defensegrid:awakening');

        if($this->isPublic()) {
            $generalData = $this->xmlData->stats->general;

            $this->bronzeMedals          = (int) $generalData->bronze_medals_won->value;
            $this->silverMedals          = (int) $generalData->silver_medals_won->value;
            $this->goldMedals            = (int) $generalData->gold_medals_won->value;
            $this->levelsPlayed          = (int) $generalData->levels_played_total->value;
            $this->levelsPlayedCampaign  = (int) $generalData->levels_played_campaign->value;
            $this->levelsPlayedChallenge = (int) $generalData->levels_played_challenge->value;
            $this->levelsWon             = (int) $generalData->levels_won_total->value;
            $this->levelsWonCampaign     = (int) $generalData->levels_won_campaign->value;
            $this->levelsWonChallenge    = (int) $generalData->levels_won_challenge->value;
            $this->encountered           = (int) $generalData->total_aliens_encountered->value;
            $this->killed                = (int) $generalData->total_aliens_killed->value;
            $this->killedCampaign        = (int) $generalData->total_aliens_killed_campaign->value;
            $this->killedChallenge       = (int) $generalData->total_aliens_killed_challenge->value;
            $this->resources             = (int) $generalData->resources_recovered->value;
            $this->heatDamage            = (float) $generalData->heatdamage->value;
            $this->timePlayed            = (float) $generalData->time_played->value;
            $this->interest              = (float) $generalData->interest_gained->value;
            $this->damage                = (float) $generalData->tower_damage_total->value;
            $this->damageCampaign        = (float) $generalData->tower_damage_total_campaign->value;
            $this->damageChallenge       = (float) $generalData->tower_damage_total_challenge->value;
            $this->orbitalLaserFired     = (int) $this->xmlData->stats->orbitallaser->fired->value;
            $this->orbitalLaserDamage    = (float) $this->xmlData->stats->orbitallaser->damage->value;
        }
    }

    /**
     * Returns stats about the towers built
     *
     * The array returned uses the names of the aliens as keys. Every value of
     * the array is an array containing the number of aliens encountered as the
     * first element and the number of aliens killed as the second element.
     */
    public function getAlienStats() {
        if(!$this->isPublic()) {
            return;
        }

        if(empty($this->alienStats)) {
            $alienData = $this->xmlData->stats->aliens;
            $this->alienStats = array();
            $aliens = array('bulwark', 'crasher', 'dart', 'decoy', 'drone',
                'grunt', 'juggernaut', 'manta', 'racer', 'rumbler', 'seeker',
                'spire', 'stealth', 'swarmer', 'turtle', 'walker');

            foreach($aliens as $alien) {
                $this->alienStats[$alien] = array(
                    (int) $alienData->$alien->encountered->value,
                    (int) $alienData->$alien->killed->value
                );
            }
        }

        return $this->alienStats;
    }

    /**
     * Returns stats about the towers built
     *
     * The Hash returned uses the names of the towers as keys. Every value of
     * the Hash is another Hash using the keys 1 to 3 for different tower
     * levels.
     * The values of these Hash is an Array containing the number of towers
     * built as the first element and the damage dealt by this specific tower
     * type as the second element.
     *
     * The Command tower uses the resources gained as second element.
     * The Temporal tower doesn't have a second element.
     *
     * @return array
     */
    public function getTowerStats() {
        if(!$this->isPublic()) {
            return;
        }

        if(empty($this->towerStats)) {
            $towerData = $this->xmlData->stats->towers;
            $this->towerStats = array();
            $towers = array('cannon', 'flak', 'gun', 'inferno', 'laser',
                'meteor', 'missile', 'tesla');

            foreach($towers as $tower) {
                $this->towerStats[$tower] = array();
                for($i = 1; $i <= 3; $i++) {
                    $built = $towerData->xpath("{$tower}[@level=$i]/built/value");
                    $damage = $towerData->xpath("{$tower}[@level=$i]/damage/value");
                    $this->towerStats[$tower][$i] = array(
                        (int) $built[0],
                        (float) $damage[0]
                    );
                }
            }

            $this->towerStats['command'] = array();
            for($i = 1; $i <= 3; $i++) {
                $built = $towerData->xpath("command[@level=$i]/built/value");
                $resources = $towerData->xpath("command[@level=$i]/resource/value");
                $this->towerStats['command'][$i] = array(
                    (int) $built[0],
                    (float) $resources[0]
                );
            }

            $this->towerStats['temporal'] = array();
            for($i = 1; $i <= 3; $i++) {
                $built = $towerData->xpath("temporal[@level=$i]/built/value");
                $this->towerStats['temporal'][$i] = array(
                    (int) $built[0]
                );
            }
        }

        return $this->towerStats;
    }
}
?>
